community/kernel/src/main/java/org/neo4j/kernel/impl/factory/DataSourceModule.java

```diff
@@ -0,0 +1,391 @@
+            }
+        };
+
+        this.queryExecutor = new Supplier<QueryExecutionEngine>()
+        {
+            @Override
+            public QueryExecutionEngine get()
+            {
+                return queryExecutor.get();
+            }
+        };
+    }
+
+    protected RelationshipProxy.RelationshipActions createRelationshipActions(final GraphDatabaseService graphDatabaseService,
+                                                                              final ThreadToStatementContextBridge threadToStatementContextBridge,
+                                                                              final NodeManager nodeManager, final RelationshipTypeTokenHolder relationshipTypeTokenHolder)
+    {
+        return new RelationshipProxy.RelationshipActions()
+        {
+            @Override
+            public GraphDatabaseService getGraphDatabaseService()
+            {
+                return graphDatabaseService;
+            }
+
+            @Override
+            public void failTransaction()
+            {
+                threadToStatementContextBridge.getKernelTransactionBoundToThisThread( true ).failure();
+            }
+
+            @Override
+            public void assertInUnterminatedTransaction()
+            {
+                threadToStatementContextBridge.assertInUnterminatedTransaction();
+            }
+
+            @Override
+            public Statement statement()
+            {
+                return threadToStatementContextBridge.get();
+            }
+
+            @Override
+            public Node newNodeProxy( long nodeId )
+            {
+                // only used by relationship already checked as valid in cache
+                return nodeManager.newNodeProxyById( nodeId );
+            }
+
+            @Override
+            public RelationshipType getRelationshipTypeById( int type )
+            {
+                try
+                {
+                    return relationshipTypeTokenHolder.getTokenById( type );
+                }
+                catch ( TokenNotFoundException e )
+                {
+                    throw new NotFoundException( e );
+                }
+            }
+        };
+    }
+
+    protected NodeProxy.NodeActions createNodeActions(final GraphDatabaseService graphDatabaseService,
+                                                                                  final ThreadToStatementContextBridge threadToStatementContextBridge,
+                                                                                  final NodeManager nodeManager)
+    {
+        return new NodeProxy.NodeActions()
+        {
+            @Override
+            public Statement statement()
+            {
+                return threadToStatementContextBridge.get();
+            }
+
+            @Override
+            public GraphDatabaseService getGraphDatabase()
+            {
+                // TODO This should be wrapped as well
```

Comment:
Most of this is copied verbatim from IAGDB. If we have issues with IAGDB code, then lets deal with that after this is merged.